### PR TITLE
Ports TG's Tesla Engine

### PR DESCRIPTION
**Pull request by Fox-McCloud**
re-upload of: https://github.com/ParadiseSS13/Paradise/pull/2979 

Fox, presiding over the creation and observation of the Tesla energy ball:
zap

Ports TG's Tesla engine.

Problems with the previous PR:
- Energy Balls locking up processing
- Energy ball escaping containment

Solved in this PR:
- Both of the above!

Changes from TG (all map related):

- The coils, grounding rods, and energy ball generator all were not anchored at round start, meaning you had to spend considerable time to get them all set up; on this implementation everything starts off anchored by default; only thing you have to do is decide which engine you want to use and unanchor the one you're not going to use and place it to the side, set up containment, and go about your businses.

### CODE REVIEW

**Comment by orelbon** [comment]
yes pls :+1:

**Comment by tigercat2000** [comment]
Please don't leave MetaStation uneditted :crying_cat_face:

File: code/modules/power/singularity/generator.dm [hunk 1]
```diff
@@ -15,7 +16,7 @@
 		message_admins("A singularity has been created at [x], [y], [z] (<A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[x];Y=[y];Z=[z]'>JMP</a>)")
```
Comment (tigercat2000) [inline code/modules/power/singularity/generator.dm]: should change `singularity` to `[creation_type]`

File: code/modules/power/tesla/energy_ball.dm [hunk 2]
```diff
@@ -0,0 +1,203 @@
+#define TESLA_DEFAULT_POWER 3476520
+#define TESLA_MINI_POWER 1738260
+
+var/list/blacklisted_tesla_types = list(/obj/machinery/atmospherics,
+										/obj/machinery/power/emitter,
+										/obj/machinery/field/generator,
+										/mob/living/simple_animal,
```
Comment (tigercat2000) [inline code/modules/power/tesla/energy_ball.dm]: No fried corgi steak?

**Comment by Fox-McCloud** [comment]
>Please don't leave MetaStation uneditted

Considering the massive amount of edits/errors that have to be corrected, I'd rather it be added when those are addressed.

**Comment by FalseIncarnate** [comment]
:+1: for new power source

Not sure I care for the removal of the Siemens Coefficient defines from all the clothing that you adjusted, but I suppose it's effects were rather small and thus makes it hard to tell how much this buffs things like Chain Lightning and Power Gloves.

**Comment by Fox-McCloud** [comment]
Now that halloss tasers are dead, the siemens coefficient change doesn't really impact much (that's purely why Bay did what they did...no idea why they didn't do the sane thing and go by energy armor...).

The only two electrocution types that went by something other than the gloves you're wearing are the wizard spell for shocks and the the ascendant shadowling shock spell---everything else literally goes by the gloves you're wearing on your hands---so, and ever so slight buff to wizard, but in practicality, it's of no impact.

**Comment by tigercat2000** [comment]
>Considering the massive amount of edits/errors that have to be corrected, I'd rather it be added when those are addressed.

Most of them are fairly minor "get around to it" things, and it makes it more of a pain in the ass when I do get to fixing it when "figure out how the tesla engine works, add all of it's components" is part of the todo list

**Comment by Fox-McCloud** [comment]
ported over the fingerprint transfer and implemented the singularity type created.

**Comment by Iamgoofball** [comment]
I've got a rework pr for when this is ready

Also, you can't have them anchored via map maker, that doesn't call connect_to_network() on it

File: code/modules/power/singularity/generator.dm [hunk 3]
```diff
@@ -8,14 +8,16 @@
...
 /obj/machinery/the_singularitygen/process()
 	var/turf/T = get_turf(src)
 	if(src.energy >= 200)
-		message_admins("A singularity has been created at [x], [y], [z] (<A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[x];Y=[y];Z=[z]'>JMP</a>)")
-		investigate_log("A singularity has been created at [x], [y], [z]","singulo")
+		message_admins("A [creation_type.name] has been created at [x], [y], [z] (<A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[x];Y=[y];Z=[z]'>JMP</a>)")
+		investigate_log("A [creation_type.name] has been created at [x], [y], [z]","singulo")
 
-		new /obj/singularity/(T, 50)
+ 		var/obj/singularity/S = new creation_type(T, 50)
```
Comment (FalseIncarnate) [inline code/modules/power/singularity/generator.dm]: This isn't playing nice with TRAVIS, it thinks you are trying to make a new proc.

Originally, this was /obj/singularity/(T, 50) which presumably is not the same as /obj/singularity(T,50) as this line would result in.

Comment (Iamgoofball) [inline code/modules/power/singularity/generator.dm]: This should work, dunno why it isn't

Comment (Fox-McCloud) [inline code/modules/power/singularity/generator.dm]: it's a single errant space.

**Comment by Fox-McCloud** [comment]
a single errant space was preventing that from going through (and a derp on my part).

**Comment by Deanthelis** [comment]
I always approve of new engines, so long as they work, and this one seems to work, so. Whoopie!

File: code/modules/power/tesla/energy_ball.dm (hunk 2 above)
Comment (Iamgoofball) [inline code/modules/power/tesla/energy_ball.dm]: simple animals + tesla = bad times

**Comment by Earthdivine** [comment]
Snapshots of the map change please?

**Comment by Fox-McCloud** [comment]
@Earthdivine 

Included photos of the map change--it's minimal, at best; just some coils and grounding rods in the singularity area+platforms to accommodate them+the tables.

some warning stripes added to the singulo pad+near emitters.

**Comment by orelbon** [comment]
god damn it fox dont be lazy and put the warning stripe corners.

**Comment by Deanthelis** [comment]
So, let me try to summarize this engine in simple terms.

You turn containment on the same way you would for the singulo. Emitters and field generators.
Make sure you wrench down the grounding rods so the electrobolts don't zap everything on the station. Or, as a traitor, unwrench them and leave the PA on maximum! Observe everyone die!
You put together the PA and turn it on as normal. Let it run until the big energy ball shows up. Continue running it until some more mini-energy balls show up and start orbiting the big one. The grounding rods, as long as they aren't sabotaged, will prevent the power from going anywhere bad and murderifying people (?).
(Is it possible to just leave the PA on, or better yet, hack it and raise it to 3 to make EVEN MORE ENERGY BAWLS?)
Walk away, engine will now produce power forever, grounding rods don't need a wire or a powernet so you're safe forever.

**Comment by Fox-McCloud** [comment]
You can leave the PA on max power 24/7 with no ill effects--it just generates more and more energy balls.

Nothing need to be wrenched down; you just unwrench the singulo or centra tesla coil and move the one to the side you don't want to use---set up is identical---containment then firing of the PA.

The grounding rods absorb the shocks (which bounce around). If you remove them, then it'll start shocking people in engineering.

The only way for this to get out is if it's DELIBERATELY let out.

No, you can't be a tesla coil.

**Comment by Deanthelis** [comment]
Great, I'm going to try a duration test. I'll set it up locally and leave it running for a full two hours, see what happens, if it's horribly laggy, etc.

**Comment by Fox-McCloud** [comment]
>if it's horribly laggy, etc.

It will be--if you max out the PA, it can easily reach into the hundreds of energy balls.

Surprisingly, the biggest cost to this is the beams it shoots and not so much the orbiting balls. 

I'll be addressing this very soon---TG also has problems with this (even more so when engineers bought 3 PA's and maxed them all out for over an hour...)

**Comment by Iamgoofball** [comment]
i highly reccomend porting in the Orbit and SpinAnimation changes again, they help with any potential lag from the teslaball orbits AND make it look cooler

**Comment by Deanthelis** [comment]
I'm 30 minutes into my test.
I've got two PA's up and running, and a huge cluster of energy balls firing away, but for whatever reason, I can't break 6 million watts.
Is this intentional?

**Comment by Fox-McCloud** [comment]
@Deanthelis 

Yeah, it tapers off; when something is being shocked, it can't be shocked by something else for a while--thus, once all of the tesla coils are being activated at once (or within sconds of each other), you've hit peak power output.

That said, you can upgrade the capacitors inside the coils to increase power output DRAMATICALLY.

9 coils being hit at once with tier-4 parts is around 19-20,000,000 watts of power.

**Comment by Fox-McCloud** [comment]
@Earthdivine 

It lives again!

**Comment by tigercat2000** [comment]
@Fox-McCloud I still want MetaStation updated, as *more* issues shouldn't be added, when you know how the damn system works, and could probably do the mapping in 5 minutes, as opposed to me spending my time figuring out how it works, trying to read over the text diff of this PR, and applying it, plus figuring out how to actually test it.

**Comment by Fox-McCloud** [comment]
Implemented Goof's Tesla changes.

Ported over TG's orbit() and SpinAnimation update. I know, @tigercat2000; it's more complicated and difficult to use, but the performance savings are *gigantic*. Orbit() is actually incredibly expensive, mostly because it's a while loop that calls animate() every few deciseconds until the end of time.

how gigantic? I'm talking the CPU of orbit() being at something like 55.0 with 60 tesla balls vs something like 0.3 with Remie's new orbit().

Compare this to @RemieRichards new orbit changes and it's way different; it makes a limited number of animate() calls and that's it; it's done and over with---this not only saves on a TON of server resources (remember how much animte() crushed us for lighting and blob's pulsing animation?), but it makes things far easier for clients to handle.

**Comment by RemieRichards** [comment]
The complexity is only for the extra features I accidentally (yes, Accidentally) implemented when rewriting ```orbit()```, the ability to produce various non-circle shapes (Triangles, Squares, Pentagon, Hexagon, Octagon, etc.)

```orbit()``` defaults to a circular orbit however, so for basic uses it's the same, just some moved arguments (that you probably won't need to touch)

Complexity should go up again soon as I have an update written that allows for even more exotic shapes such as pentaGRAMS and triskaidecagons (13 pointed stars), but to alleviate that extra complexity I'll produce simple #defines that contain the necessary instructions to build the simpler shapes, eg: ```#define ORBIT_SHAPE_PENTAGRAM list(144,144,144,144,144)``` (yes that's the actual instructions) 

pentagram

triskaidecagon

but yes, tl;dr new ```orbit()``` is INFINITELY cheaper then my previous attempt, I really outdid myself :P

**Comment by Earthdivine** [comment]
@Fox-McCloud I would really prefer if it got put onto Meta now rather then later, that could cause complications down the line with other Pull's. Plus it's a good time waster waiting on other things to go through, ya?

**Comment by Fox-McCloud** [comment]
@Earthdivine 

Just updated for that.

@RemieRichards 

Yeah; thanks for updating orbit and optimizing the shit out of it; if BYOND were sane and animate() was handled completely clientside, it would be a neat little thing to do, but since it literally renders every frame, server side, then pushes it out to all clients, it's *insanely* expensive, especially when it's a loop of some kind.

**Comment by CrAzYPiLoT-SS13** [comment]
MERGE!

**Comment by Taaketa** [comment]
Please wanna try it :) +1:

**Comment by Deanthelis** [comment]
Having tested it locally and produced the results Fox talked about a bit ago, and having tested what can happen if you let it go, I throw my full, tiny amount of support behind this, especially now that it's mapped into Metastation appropriately and everything seems to be worked out.

Fox, I do think the Orbit() changes should've been in another PR, because isn't Orbit() called elsewhere as well? What else would be affected by these changes besides the energy balls?

Also, there is one thing I did notice, but forgot to report. When aghosted/ghosted and using Observe, the main energy ball appears as 'Gravitational Singularity' and EVERY SINGLE mini-ball appears as Gravitational Singularity (1), (2), (3), etc., which is extremely annoying when you have, say, 50.

**Comment by tigercat2000** [comment]
@Deanthelis Orbit isn't used anywhere else, actually, as far as I know.

**Comment by Deanthelis** [comment]
Huh.

Then, why all the fuss about it? I don't understand.

**Comment by RemieRichards** [comment]
@Deanthelis it's a feature you guys chose to skip out on due to performance issues, there are now no performance issues -at all- from using it, so you don't have to miss out any more.

that and goof used it in the tesla engine, so you'd have to port the improvements OR miss out on that feature of the tesla.

**Comment by tigercat2000** [comment]
Actually @RemieRichards, I ported it when it was first made, then subsequently never updated it or used it because it became much more complicated for each use on -tg-.

**Comment by RemieRichards** [comment]
That's not what I recall at all, I remember you guys having issues with animate due to your lighting system ALSO over-using it, preventing you from really having both running side-by-side.

MrStonedOne (the only person to touch orbit after the initial PR by myself) only did bugfixes and added 1 extra argument to the proc and I don't know how 1 extra argument is any major increased level of complexity.

The new version defaults to values akin to the ease of use of the first version, any extra arguments are either a new syntax for producing shapes or extra functionality you don't *have* to touch.

tl;dr this is not a complex function, I'm very confused as to why you think it is, the new version even comes with more documentation.

**Comment by tigercat2000** [comment]
@RemieRichards We did have pretty big issues with animate() from the lighting system, but as far as I know, the only thing we even talked about with the orbit system was ghosts orbiting, and never actually made any work on it because of worrying over animate().

Doesn't really matter though, I suppose it's only as complicated as you think it is, and the code controlling it does look more complicated than it probably actually is.

**Comment by Fox-McCloud** [comment]
@Deanthelis 

While I don't have to port Orbit() in this PR, what's going to end up happening is this, if I don't:

- Shift is at nearly 2 hours in and there's something like just shy of 200 energy balls orbiting the primary energy ball---the lag generates from all the orbit() calls will be insane and astronoimcal---not to mention people will just plain crash when they get close to it.

**Comment by Fox-McCloud** [comment]
@Markolie  @TheDZD

**Comment by orelbon** [comment]
so what would happen if say you activated the singulo and the tesla engine at the same time?

**Comment by Fox-McCloud** [comment]
When the singulo inevitably bumps into the tesla ball---there's going to be an explosion....a very big one.

**Comment by orelbon** [comment]
i know what i'm gonna accidentally do when this gets patched....

**Comment by Iamgoofball** [comment]
Warranty void if orbit changes aren't ported.

**Comment by Fox-McCloud** [comment]
@Iamgoofball 

https://github.com/Fox-McCloud/Paradise-1/commit/bfb23936f9f505948ae79b574045a10cf9b73c24

**Comment by CrAzYPiLoT-SS13** [comment]
YES!

**Comment by orelbon** [comment]
sooo now we just need a patch. whenever that may come...

**Comment by Plotron** [comment]
Regarding that crashing thing...
I joined as a spectator, turned on Antag HUD and Med HUD, someone asked a question about Tesla in deadchat, so I teleported to him. I only caught a glimpse of the engine containment and surroundings, and I crashed. Rejoining always results in a crash. I'm glad I wasn't playing as AI!

**Comment by Robertha89** [comment]
Oh, that's heap corruption, you got two choices, rapid re-connect untill it does not crash, and it can take any from 1 to a 100 or more, or ask an admin to move your character somewhere else, apparently this is caused by byond beta, I believe the lighting system to be the source tbh.

**Comment by Spacemanspark** [comment]
That's why you use v508.1292.

**Comment by Crazylemon64** [comment]
The new version's graphics have inconsistently-sized pixels for me, which makes everything look super ugly. Any advice?

**Comment by tigercat2000** [comment]
@Crazylemon64 It's probably BYOND's automatic scaling for your screen, have to either hit the "icons" dropdown and select `64x64`/`32x32` or live with it.

**Comment by Spacemanspark** [comment]
...Am I the only one NOT getting the "you must upgrade BYOND to play" message?

**Comment by Crazylemon64** [comment]
Ah, nah, my current version works fine, and I'm using it as long as I can get away with it. Compiling needs the new version, but I've got a separate pack of that handy.
